File: app/mac/Sources/Theme.swift
// Solarized. Light and dark share the accents; only the eight greys swap,
// which is the whole point of the palette.
import AppKit
import SwiftUI

enum Sol {
    static let base03 = NSColor(srgbRed: 0x00 / 255, green: 0x2b / 255, blue: 0x36 / 255, alpha: 1)
    static let base02 = NSColor(srgbRed: 0x07 / 255, green: 0x36 / 255, blue: 0x42 / 255, alpha: 1)
    static let base01 = NSColor(srgbRed: 0x58 / 255, green: 0x6e / 255, blue: 0x75 / 255, alpha: 1)
    static let base00 = NSColor(srgbRed: 0x65 / 255, green: 0x7b / 255, blue: 0x83 / 255, alpha: 1)
    static let base0  = NSColor(srgbRed: 0x83 / 255, green: 0x94 / 255, blue: 0x96 / 255, alpha: 1)
    static let base1  = NSColor(srgbRed: 0x93 / 255, green: 0xa1 / 255, blue: 0xa1 / 255, alpha: 1)
    static let base2  = NSColor(srgbRed: 0xee / 255, green: 0xe8 / 255, blue: 0xd5 / 255, alpha: 1)
    static let base3  = NSColor(srgbRed: 0xfd / 255, green: 0xf6 / 255, blue: 0xe3 / 255, alpha: 1)

    static let yellow  = Color(nsColor: NSColor(srgbRed: 0xb5 / 255, green: 0x89 / 255, blue: 0x00 / 255, alpha: 1))
    static let orange  = Color(nsColor: NSColor(srgbRed: 0xcb / 255, green: 0x4b / 255, blue: 0x16 / 255, alpha: 1))
    static let red     = Color(nsColor: NSColor(srgbRed: 0xdc / 255, green: 0x32 / 255, blue: 0x2f / 255, alpha: 1))
    static let magenta = Color(nsColor: NSColor(srgbRed: 0xd3 / 255, green: 0x36 / 255, blue: 0x82 / 255, alpha: 1))
    static let violet  = Color(nsColor: NSColor(srgbRed: 0x6c / 255, green: 0x71 / 255, blue: 0xc4 / 255, alpha: 1))
    static let blue    = Color(nsColor: NSColor(srgbRed: 0x26 / 255, green: 0x8b / 255, blue: 0xd2 / 255, alpha: 1))
    static let cyan    = Color(nsColor: NSColor(srgbRed: 0x2a / 255, green: 0xa1 / 255, blue: 0x98 / 255, alpha: 1))
    static let green   = Color(nsColor: NSColor(srgbRed: 0x85 / 255, green: 0x99 / 255, blue: 0x00 / 255, alpha: 1))

    /// Follows the OS, without every view having to know which mode it is in.
    private static func dyn(_ light: NSColor, _ dark: NSColor) -> Color {
        Color(nsColor: NSColor(name: nil) { appearance in
            appearance.bestMatch(from: [.aqua, .darkAqua]) == .darkAqua ? dark : light
        })
    }

    static let bg     = dyn(base3, base03)
    static let bgAlt  = dyn(base2, base02)
    static let fg     = dyn(base00, base0)
    static let fgEm   = dyn(base01, base1)
    static let fgDim  = dyn(base1, base01)
    static let rule   = dyn(base01.withAlphaComponent(0.18), base1.withAlphaComponent(0.16))
    static let onAccent = dyn(base3, base03)

    /// One stable colour per person, so you can tell who is who at a glance.
    /// Someone and their AI share a colour — it is the same household.
    static func forName(_ name: String) -> Color {
        let palette: [Color] = [blue, magenta, cyan, violet, orange, green, yellow]
        var h: UInt32 = 0
        for b in name.utf8 { h = h &* 31 &+ UInt32(b) }
        return palette[Int(h % UInt32(palette.count))]
    }

    static func forAction(_ action: String) -> Color {
        switch action {
        case "added": return green
        case "removed": return red
        case "renamed": return violet
        default: return blue
        }
    }
}

File: app/windows/Theme.cs
// Solarized, the same palette the Mac app uses. Light and dark share the eight
// accents; only the greys swap, which is the point of the scheme.
using System;
using System.Windows.Media;

namespace Collab
{
    public static class Sol
    {
        static SolidColorBrush B(byte r, byte g, byte b)
        {
            var s = new SolidColorBrush(Color.FromRgb(r, g, b));
            s.Freeze();
            return s;
        }

        public static readonly SolidColorBrush Base03 = B(0x00, 0x2b, 0x36);
        public static readonly SolidColorBrush Base02 = B(0x07, 0x36, 0x42);
        public static readonly SolidColorBrush Base01 = B(0x58, 0x6e, 0x75);
        public static readonly SolidColorBrush Base00 = B(0x65, 0x7b, 0x83);
        public static readonly SolidColorBrush Base0  = B(0x83, 0x94, 0x96);
        public static readonly SolidColorBrush Base1  = B(0x93, 0xa1, 0xa1);
        public static readonly SolidColorBrush Base2  = B(0xee, 0xe8, 0xd5);
        public static readonly SolidColorBrush Base3  = B(0xfd, 0xf6, 0xe3);

        public static readonly SolidColorBrush Yellow  = B(0xb5, 0x89, 0x00);
        public static readonly SolidColorBrush Orange  = B(0xcb, 0x4b, 0x16);
        public static readonly SolidColorBrush Red     = B(0xdc, 0x32, 0x2f);
        public static readonly SolidColorBrush Magenta = B(0xd3, 0x36, 0x82);
        public static readonly SolidColorBrush Violet  = B(0x6c, 0x71, 0xc4);
        public static readonly SolidColorBrush Blue    = B(0x26, 0x8b, 0xd2);
        public static readonly SolidColorBrush Cyan    = B(0x2a, 0xa1, 0x98);
        public static readonly SolidColorBrush Green   = B(0x85, 0x99, 0x00);

        /// Set once at startup and whenever Windows switches mode.
        public static bool Dark { get; set; } = true;

        public static SolidColorBrush Bg       => Dark ? Base03 : Base3;
        public static SolidColorBrush BgAlt    => Dark ? Base02 : Base2;
        public static SolidColorBrush Fg       => Dark ? Base0  : Base00;
        public static SolidColorBrush FgEm     => Dark ? Base1  : Base01;
        public static SolidColorBrush FgDim    => Dark ? Base01 : Base1;
        public static SolidColorBrush OnAccent => Dark ? Base03 : Base3;

        public static SolidColorBrush Rule
        {
            get
            {
                var c = (Dark ? Base1 : Base01).Color;
                var s = new SolidColorBrush(Color.FromArgb(Dark ? (byte)41 : (byte)46, c.R, c.G, c.B));
                s.Freeze();
                return s;
            }
        }

        /// One stable colour per person. The hash matches the Mac's exactly, so
        /// the same name is the same colour on both machines — otherwise two
        /// people looking at one conversation would be reading different
        /// pictures of it.
        public static SolidColorBrush ForName(string name)
        {
            var palette = new[] { Blue, Magenta, Cyan, Violet, Orange, Green, Yellow };
            uint h = 0;
            foreach (var b in System.Text.Encoding.UTF8.GetBytes(name)) h = unchecked(h * 31 + b);
            return palette[(int)(h % (uint)palette.Length)];
        }

        public static SolidColorBrush ForAction(string action) => action switch
        {
            "added"   => Green,
            "removed" => Red,
            "renamed" => Violet,
            _         => Blue,
        };

        /// Windows records the choice in the registry; there is no notification
        /// worth the plumbing, so this is read at startup and on window focus.
        public static bool SystemPrefersDark()
        {
            try
            {
                using var k = Microsoft.Win32.Registry.CurrentUser.OpenSubKey(
                    @"Software\Microsoft\Windows\CurrentVersion\Themes\Personalize");
                return (k?.GetValue("AppsUseLightTheme") as int?) == 0;
            }
            catch { return true; }
        }
    }
}
